zio-http/src/main/scala/zhttp/http/Middleware.scala

```diff
@@ -1,450 +1,301 @@
-
-  /**
-   * Creates a new middleware using a function from request parameters to a ZIO of HttpMiddleware
-   */
-  def fromMiddlewareFunctionZIO[R, E](f: RequestP[ZIO[R, Option[E], Middleware[R, E]]]): Middleware[R, E] =
-    Middleware.FromFunctionZIO(f)
-
-  /**
-   * An empty middleware that doesn't do anything
-   */
-  def identity: Middleware[Any, Nothing] = Identity
+  final def orElse[R1 <: R, E1, AIn0 >: AIn, BIn0 <: BIn, AOut0 <: AOut, BOut0 >: BOut](
+    other: Middleware[R1, E1, AIn0, BIn0, AOut0, BOut0],
+  ): Middleware[R1, E1, AIn0, BIn0, AOut0, BOut0] =
+    Middleware.OrElse(self, other)
 
   /**
-   * Logical operator to decide which middleware to select based on the header
+   * Race between current and other, cancels other when execution of one completes
    */
-  def ifHeader[R, E](cond: Headers => Boolean)(left: Middleware[R, E], right: Middleware[R, E]): Middleware[R, E] =
-    ifThenElse((_, _, headers) => cond(headers))(left, right)
-
-  /**
-   * Logical operator to decide which middleware to select based on the predicate.
-   */
-  def ifThenElse[R, E](
-    cond: RequestP[Boolean],
-  )(left: Middleware[R, E], right: Middleware[R, E]): Middleware[R, E] =
-    Middleware.FromFunctionZIO((method, url, headers) => UIO(if (cond(method, url, headers)) left else right))
+  final def race[R1 <: R, E1 >: E, AIn1 >: AIn, BIn1 <: BIn, AOut1 <: AOut, BOut1 >: BOut](
+    other: Middleware[R1, E1, AIn1, BIn1, AOut1, BOut1],
+  ): Middleware[R1, E1, AIn1, BIn1, AOut1, BOut1] =
+    Middleware.Race(self, other)
 
   /**
-   * Logical operator to decide which middleware to select based on the predicate.
+   * Applies Middleware based only if the condition function evaluates to true
    */
-  def ifThenElseZIO[R, E](
-    cond: RequestP[ZIO[R, E, Boolean]],
-  )(left: Middleware[R, E], right: Middleware[R, E]): Middleware[R, E] =
-    Middleware.FromFunctionZIO((method, url, headers) =>
-      cond(method, url, headers).mapBoth(
-        Option(_),
-        {
-          case true  => left
-          case false => right
-        },
-      ),
+  final def when[AOut0 <: AOut](cond: AOut0 => Boolean): Middleware[R, E, AIn, BIn, AOut0, BOut] =
+    Middleware.ifThenElse[AOut0](cond(_))(
+      isTrue = _ => self,
+      isFalse = _ => Middleware.identity,
     )
 
   /**
-   * Creates a new middleware using transformation functions
+   * Applies Middleware and returns a transformed Http app
    */
-  def make[S](req: (Method, URL, Headers) => S): PartiallyAppliedMake[S] = PartiallyAppliedMake(req)
+  private[zhttp] final def execute[R1 <: R, E1 >: E](http: Http[R1, E1, AIn, BIn]): Http[R1, E1, AOut, BOut] =
+    Middleware.execute(http, self)
+}
 
-  /**
-   * Creates a new middleware using effectful transformation functions
-   */
-  def makeZIO[R, E, S](req: (Method, URL, Headers) => ZIO[R, Option[E], S]): PartiallyAppliedMakeZIO[R, E, S] =
-    PartiallyAppliedMakeZIO(req)
+object Middleware extends HttpMiddlewares {
```

Comment:
should add a `timeout` middleware

Reply:
What should be the desired result
```scala
def timeout[B](duration: Duration): Middleware[Clock, Throwable, Nothing, Any, Any, B] =
    Middleware.identity.race(Middleware.fail(new Throwable("Time out")).delay(duration))
```

Reply:
https://github.com/dream11/zio-http/pull/798
We have discussed to not do this change